File: cfg_maker/cs_test/cross/Station.cs
/*
 * Cross Call
 */
using System;
using System.Collections.Generic;
using System.Runtime.InteropServices;

namespace CrossCall
{
    public class Station
    {
        protected class Invok : IInvoker
        {
            Station _station;

            public Invok(Station station)
            { _station = station; }
            public Serialize.IWriter Begin(uint module)
            { return _station.BeginCall(module); }
            public Serialize.IReader End()
            { return _station.EndCall(); }
        }

        protected class Context : IContext
        {
            Station _station;

            public Context(Station station)
            { _station = station; }
            public Serialize.IReader Param
            { get { return _station._reader; } }
            public Serialize.IWriter Ret()
            { return _station.RetParam(); }
        }

        ICrossCall _caller;
        IntPtr _ptr;
        int _size;

        byte[] _pool;
        Invok _invoker;
        Context _param;

        SwapBuffer _buffer;
        Serialize.BinaryReader _reader;
        Serialize.BinaryWriter _writer;

        Dictionary<uint, IProcessor> _dicProc = new Dictionary<uint, IProcessor>();

        public Station(ICrossCall caller, IntPtr ptr, int size)
        {
            _caller = caller;
            _ptr = ptr;
            _size = size;

            _pool = new byte[size];
            _invoker = new Invok(this);
            _param = new Context(this);
            _buffer = new SwapBuffer(_pool);
            _reader = new Serialize.BinaryReader(_buffer);
            _writer = new Serialize.BinaryWriter(_buffer);
        }

        public IInvoker Invoker { get { return _invoker; } }

        public bool Register(uint module, IProcessor processor)
        {
            if (_dicProc.ContainsKey(module))
                return false;

            _dicProc.Add(module, processor);
            return true;
        }

        public IProcessor GetProcessor(uint module)
        {
            IProcessor proc = null;
            _dicProc.TryGetValue(module, out proc);
            return proc;
        }

        public IProcessor Unregister(uint module)
        {
            IProcessor proc = null;
            if (_dicProc.TryGetValue(module, out proc))
                _dicProc.Remove(module);
            return proc;
        }

        public void OnCall()
        {
            DoRecv();

            uint module = 0;
            IProcessor proc = null;
            if (_reader.Read(ref module, null) && _dicProc.TryGetValue(module, out proc))
            {
                proc.Process(_param);
            }
            else
            {
                // error
            }

            DoSend();
        }

        Serialize.IWriter BeginCall(uint module)
        {
            _buffer.Startup(BufferMode.Write, 0);
            _writer.Write(module);
            return _writer;
        }

        Serialize.IReader EndCall()
        {
            DoSend();
            _caller.DoCall();
            DoRecv();
            return _reader;
        }

        Serialize.IWriter RetParam()
        {
            _buffer.Startup(BufferMode.Write, 0);
            return _writer;
        }

        void DoRecv()
        {
            int size = Marshal.ReadInt32(_ptr, 0);
            Marshal.Copy(_ptr, _pool, 0, size);
            _buffer.Startup(BufferMode.Read, size);
        }

        void DoSend()
        {
            int size = _buffer.Endup();
            if (_buffer.Mode == BufferMode.Write)
                Marshal.Copy(_pool, 0, _ptr, size);
            else
                Marshal.WriteInt32(_ptr, 0);    // clear
        }
    }
}

File: cfg_maker/dll_test/cross/Station.cpp
#include "Station.h"

namespace cross_call
{
    SwapBuffer::SwapBuffer(char* buffer, int size)
        : _header((SwapBuffer::Header*)buffer)
        , _base((char*)buffer)
        , _size(size)
    {
    }

    void SwapBuffer::Startup(BufferMode mode)
    {
        _mode = mode;
        _rp = _wp = 4;
    }

    int SwapBuffer::Endup()
    {
        if (_mode == BufferMode::Write)
            _header->ds = _wp;
        else
            _header->ds = 0;

        _rp = _wp = 4;
        return _header->ds;
    }

    bool SwapBuffer::Read(void* buf, uint32_t size)
    {
        if (_mode != BufferMode::Read)
            return false;
        if (_rp + (int)size > _header->ds)
            return false;

        memcpy(buf, _base + _rp, size);
        _rp += (int)size;
        return true;
    }

    bool SwapBuffer::Write(const void* buf, uint32_t size)
    {
        if (_mode != BufferMode::Write)
            return false;
        if (_wp + (int)size > _size)
            return false;

        memcpy(_base + _wp, buf, size);
        _wp += (int)size;
        return true;
    }

    //////////////////////////////////////////////////////////////////////////
    // Station
    Station::Station(ICrossCall* caller, char* buffer, int size)
        : _caller(caller)
        , _context(*this)
        , _cross(*this)
    {
        _buffer = new SwapBuffer(buffer, size);
        _reader = new serialize::BinaryReader(_buffer);
        _writer = new serialize::BinaryWriter(_buffer);
    }

    Station::~Station()
    {
        delete _reader;
        delete _writer;
        delete _buffer;
    }

    bool Station::Register(uint32_t module, IProcessor* processor)
    {
        return _procs.insert(std::make_pair(module, processor)).second;
    }

    IProcessor* Station::Unregister(uint32_t module)
    {
        IProcessor* proccessor = GetProcessor(module);
        if (proccessor)
            _procs.erase(module);
        return proccessor;
    }

    IProcessor* Station::GetProcessor(uint32_t module) const
    {
        const auto it = _procs.find(module);
        if (it == _procs.cend())
            return nullptr;
        return it->second;
    }

    void Station::OnCall()
    {
        uint32_t module = 0;
        IProcessor* proc = nullptr;

        _buffer->Startup(BufferMode::Read);
        _reader->Read(module);
        proc = GetProcessor(module);

        if (proc)
        {
            proc->Process(&_context);
        }
        else
        {
            //TODO: error
        }

        _buffer->Endup();
    }

    serialize::IWriter* Station::RetParam()
    {
        _buffer->Startup(BufferMode::Write);
        return _writer;
    }

    serialize::IWriter* Station::BeginCall(uint32_t module)
    {
        _buffer->Startup(BufferMode::Write);
        _writer->Write(module);
        return _writer;
    }

    serialize::IReader* Station::EndCall()
    {
        _buffer->Endup();

        _caller->DoCall();
        _buffer->Startup(BufferMode::Read);
        return _reader;
    }
}
